projects/plugins/jetpack/tests/php/modules/subscriptions/test_class.jetpack-subscriptions.php

```diff
@@ -59,20 +57,307 @@ protected function setup_jetpack_paid_newsletters( $subscription_end_date = null
+			$this->accessUseCase( 'paid_subscriber_id', true, false, 'paid_subscribers', false, false, $time_outdated ),
+			// token
+			$this->accessUseCase( 'paid_subscriber_id', true, false, '', true, true, $time_outdated ),
+			$this->accessUseCase( 'paid_subscriber_id', false, true, 'everybody', true, true, $time_outdated ),
+			$this->accessUseCase( 'paid_subscriber_id', false, true, 'subscribers', true, true, $time_outdated ),
+			$this->accessUseCase( 'paid_subscriber_id', false, true, 'paid_subscribers', false, false, $time_outdated ),
 
-		// Fake subscription for the paid user
+			// inactive subscription status
+			$this->accessUseCase( 'paid_subscriber_id', true, false, 'paid_subscribers', false, false, null, 'inactive' ),
+
+		);
+	}
+
+	/**
+	 * Stubs WPCOM_Token_Subscription_Service in order to return the provided token.
+	 *
+	 * @param array $payload
+	 * @return mixed
+	 */
+	private function setReturnedToken( $payload ) {
+		// We remove anything else
+		remove_all_filters( 'earn_get_user_subscriptions_for_site_id' );
+		$service       = new WPCOM_Token_Subscription_Service();
+		$_GET['token'] = JWT::encode( $payload, $service->get_key() );
+		return $service;
+	}
+
+	/**
+	 * Retrieves payload for JWT token
+	 *
+	 * @param bool $is_subscribed
+	 * @param bool $is_paid_subscriber
+	 * @param int  $subscription_end_date
+	 * @return array
+	 */
+	private function getPayload( $is_subscribed, $is_paid_subscriber, $subscription_end_date, $status ) {
+		$subscriptions = ! $is_paid_subscriber ? array() : array(
+			$this->product_id => array(
+				'status'     => $status ? $status : 'active',
+				'end_date'   => $subscription_end_date ? $subscription_end_date : time() + HOUR_IN_SECONDS,
+				'product_id' => $this->product_id,
+			),
+		);
+
+		return array(
+			'blog_sub'      => $is_subscribed ? 'active' : 'inactive',
+			'subscriptions' => $subscriptions,
+		);
+	}
+
+	/**
+	 * Test the whole matrix access
+	 *
+	 * @dataProvider matrixAccess
+	 */
+	public function testSubscriberAccessLevel( $type_user_id, $logged, $token_set, $post_access_level, $should_email_be_sent, $should_user_access_post, $subscription_end_date = null, $status = null ) {
+		if ( $type_user_id !== null ) {
+			$user_id = $this->{$type_user_id};
+		} else {
+			$user_id = 0;
+		}
+
+		$is_blog_subscriber = $user_id === $this->paid_subscriber_id || $user_id === $this->regular_subscriber_id;
+		$is_paid_subscriber = $user_id === $this->paid_subscriber_id;
+		$payload            = $this->getPayload( $is_blog_subscriber, $is_paid_subscriber, $subscription_end_date, $status );
+
+		$post_id = $this->setup_jetpack_paid_newsletters();
+		$this->setReturnedSubscriptions( $payload );
+
+		$GLOBALS['post'] = get_post( $post_id );
+		update_post_meta( $post_id, '_jetpack_newsletter_access', $post_access_level );
+
+		$subscription_service = new WPCOM_Online_Subscription_Service();
```

Comment:
We only really check this.

Reply:
What do you mean? We check [WPCOM_Token_Subscription_Service and WPCOM_Online_Subscription_Service as well](https://github.com/Automattic/jetpack/pull/27564/files#diff-307b85ed7ed565b3adafed7b241dfeda61a1e6c228480c869c8ddd9d5c77f380R304-R318)

Reply:
I see!